<a href="https://colab.research.google.com/github/maziviero27/cp4_SERS/blob/main/CP_04_SERS_ANALISE_COMPLETA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Atividade prática com datasets de energia

## Orange Data Mining, Python e Pandas

Este notebook reúne as análises dos Datasets 1 a 6 em um único arquivo. Cada seção apresenta os códigos, os resultados executados e as respostas interpretativas.

# Dataset 1 - Appliances Energy Prediction

**Situação:** uma empresa de eficiência energética está analisando o comportamento de uma residência de baixo consumo. A equipe deseja identificar períodos de consumo elevado dos eletrodomésticos e observar quais condições de temperatura e umidade estavam presentes nesses momentos.

## 1. Carregamento e inspeção inicial da amostra

In [5]:
from pathlib import Path
import pandas as pd

nomes_dataset1 = [
    'energydata_SAMPLE_01.csv'
]
pastas_busca_1 = [
    Path('/content'),
    Path.cwd(),
    Path.home() / 'Downloads',
    Path.home() / 'Documents' / 'arquivos-cp-sers'
]
arquivo_dataset1 = next(
    (
        pasta / nome
        for pasta in pastas_busca_1
        for nome in nomes_dataset1
        if (pasta / nome).exists()
    ),
    None
)

if arquivo_dataset1 is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-1.csv para o Colab antes de executar.'
    )

dados1 = pd.read_csv(arquivo_dataset1)
inicio_dataset1 = ' '.join(
    map(str, dados1.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset1 or 'meta' in inicio_dataset1:
    dados1 = pd.read_csv(arquivo_dataset1, skiprows=[1, 2])

colunas_dataset1 = [
    'Appliances', 'lights', 'T1', 'RH_1', 'T2', 'RH_2', 'T3', 'RH_3'
]
dados1 = dados1[colunas_dataset1].copy()

print(f'Dimensões da amostra: {dados1.shape[0]} linhas e {dados1.shape[1]} colunas')
dados1.head()

Dimensões da amostra: 1974 linhas e 8 colunas


,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
0,40,0,20.8900,35.4000,17.7600,39.1633,20.2900,36.9000
1,90,10,21.8900,53.1000,21.2900,45.3600,21.6333,49.2267
2,50,0,21.3900,35.5000,17.6333,40.5300,21.6667,35.2000
3,50,0,21.3900,41.0333,23.8900,34.8400,22.0333,36.9333
4,70,0,19.9633,35.1267,16.4633,40.1267,20.0000,36.4000


In [6]:
dados1.info()
print()
print('Valores ausentes por coluna:')
display(dados1.isnull().sum().to_frame('Quantidade'))
dados1.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1974 entries, 0 to 1973
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Appliances  1974 non-null   int64  
 1   lights      1974 non-null   int64  
 2   T1          1974 non-null   float64
 3   RH_1        1974 non-null   float64
 4   T2          1974 non-null   float64
 5   RH_2        1974 non-null   float64
 6   T3          1974 non-null   float64
 7   RH_3        1974 non-null   float64
dtypes: float64(6), int64(2)
memory usage: 123.5 KB

Valores ausentes por coluna:


,Quantidade
Appliances,0
lights,0
T1,0
RH_1,0
T2,0
RH_2,0
T3,0
RH_3,0


,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
count,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000,1974.000000
mean,93.044580,3.839919,21.664703,40.219193,20.293635,40.411386,22.231615,39.209733
std,93.826149,7.937076,1.574883,3.971923,2.126958,4.037686,1.957275,3.288818
min,20.000000,0.000000,16.790000,27.733300,16.100000,24.063300,17.200000,30.133300
25%,50.000000,0.000000,20.700000,37.205825,18.790000,37.760000,20.700000,36.900000
50%,60.000000,0.000000,21.600000,39.590000,20.000000,40.433300,22.100000,38.433300
75%,100.000000,0.000000,22.600000,42.991675,21.500000,43.228325,23.290000,41.741250
max,770.000000,50.000000,26.260000,57.496700,28.917500,56.026700,29.198600,49.226700


## 2. Renomeação dos atributos

In [7]:
dados1.rename(columns={
    'Appliances': 'Consumo_Eletrodomesticos',
    'lights': 'Iluminacao',
    'T1': 'Temperatura_1',
    'RH_1': 'Umidade_1',
    'T2': 'Temperatura_2',
    'RH_2': 'Umidade_2',
    'T3': 'Temperatura_3',
    'RH_3': 'Umidade_3'
}, inplace=True)

print(dados1.columns.tolist())
dados1.head()

['Consumo_Eletrodomesticos', 'Iluminacao', 'Temperatura_1', 'Umidade_1', 'Temperatura_2', 'Umidade_2', 'Temperatura_3', 'Umidade_3']


,Consumo_Eletrodomesticos,Iluminacao,Temperatura_1,Umidade_1,Temperatura_2,Umidade_2,Temperatura_3,Umidade_3
0,40,0,20.8900,35.4000,17.7600,39.1633,20.2900,36.9000
1,90,10,21.8900,53.1000,21.2900,45.3600,21.6333,49.2267
2,50,0,21.3900,35.5000,17.6333,40.5300,21.6667,35.2000
3,50,0,21.3900,41.0333,23.8900,34.8400,22.0333,36.9333
4,70,0,19.9633,35.1267,16.4633,40.1267,20.0000,36.4000


## 3. Maior consumo de eletrodomésticos

In [8]:
consumo_maximo_1 = dados1['Consumo_Eletrodomesticos'].max()
print(f'Maior consumo registrado: {consumo_maximo_1:.2f} Wh')

Maior consumo registrado: 770.00 Wh


## 4 e 5. Limiar de 70%, quantidade e percentual

In [9]:
limiar_70_1 = 0.70 * consumo_maximo_1
df_alto_consumo_1 = dados1[
    dados1['Consumo_Eletrodomesticos'] > limiar_70_1
].copy()

quantidade_alto_1 = len(df_alto_consumo_1)
percentual_alto_1 = quantidade_alto_1 / len(dados1) * 100

print(f'Limiar de 70% do máximo: {limiar_70_1:.2f} Wh')
print(f'Registros acima do limiar: {quantidade_alto_1}')
print(f'Percentual da amostra: {percentual_alto_1:.2f}%')
df_alto_consumo_1.sort_values(
    'Consumo_Eletrodomesticos', ascending=False
).head(10)

Limiar de 70% do máximo: 539.00 Wh
Registros acima do limiar: 22
Percentual da amostra: 1.11%


,Consumo_Eletrodomesticos,Iluminacao,Temperatura_1,Umidade_1,Temperatura_2,Umidade_2,Temperatura_3,Umidade_3
1733,770,20,20.2000,40.5000,19.2900,40.4333,20.5000,39.1567
916,750,30,22.4300,43.0000,21.9267,40.1333,21.2300,42.5967
744,740,20,19.5000,42.1333,18.2600,39.1967,19.8900,39.1933
1835,720,10,22.2000,37.9333,21.1000,37.0000,23.1667,36.6267
993,690,10,22.5333,45.0267,22.0667,39.7900,22.3900,41.9300
1514,690,20,21.3900,39.5900,20.8600,39.5600,23.0000,37.1633
1323,680,0,20.2900,39.3333,19.6000,38.3600,20.5333,40.3267
427,650,0,20.8900,36.7000,18.2900,39.6633,21.2000,36.0000
1964,620,0,22.7000,41.7233,22.4267,39.0000,23.2900,39.2000
352,620,0,24.3900,44.3333,25.3700,37.7360,26.7300,38.8633


## 6. Alto consumo e temperatura acima da média

In [10]:
temperatura_media_1 = dados1['Temperatura_1'].mean()
df_alto_consumo_temperatura_1 = dados1[
    (dados1['Consumo_Eletrodomesticos'] > limiar_70_1) &
    (dados1['Temperatura_1'] > temperatura_media_1)
].copy()

quantidade_duplo_1 = len(df_alto_consumo_temperatura_1)
percentual_dentro_altos_1 = quantidade_duplo_1 / quantidade_alto_1 * 100

print(f'Temperatura média de T1: {temperatura_media_1:.2f} °C')
print(f'Registros com as duas condições: {quantidade_duplo_1}')
print(f'Percentual entre os casos de alto consumo: {percentual_dentro_altos_1:.2f}%')
df_alto_consumo_temperatura_1.head(10)

Temperatura média de T1: 21.66 °C
Registros com as duas condições: 8
Percentual entre os casos de alto consumo: 36.36%


,Consumo_Eletrodomesticos,Iluminacao,Temperatura_1,Umidade_1,Temperatura_2,Umidade_2,Temperatura_3,Umidade_3
352,620,0,24.3900,44.3333,25.3700,37.7360,26.7300,38.8633
616,570,10,24.8900,32.6667,23.5600,31.1400,23.6000,30.8233
916,750,30,22.4300,43.0000,21.9267,40.1333,21.2300,42.5967
993,690,10,22.5333,45.0267,22.0667,39.7900,22.3900,41.9300
1182,600,20,24.8900,48.1333,28.9175,37.9475,27.1000,42.4400
1196,580,0,23.6333,47.2267,24.8140,43.4720,26.1667,40.5000
1835,720,10,22.2000,37.9333,21.1000,37.0000,23.1667,36.6267
1964,620,0,22.7000,41.7233,22.4267,39.0000,23.2900,39.2000


## 7. Comparação e interpretação

In [11]:
from IPython.display import Markdown, display

reducao_1 = quantidade_alto_1 - quantidade_duplo_1
reducao_percentual_1 = reducao_1 / quantidade_alto_1 * 100

display(Markdown(f'''
**Interpretação:** o filtro somente por consumo selecionou **{quantidade_alto_1} registros**. Ao incluir a temperatura acima da média, o total caiu para **{quantidade_duplo_1} registros**, redução de **{reducao_1} registros ({reducao_percentual_1:.2f}%)**.

Isso acontece porque o segundo DataFrame exige duas condições simultâneas e, portanto, contém apenas a parte dos episódios de alto consumo que também ocorreu com temperatura acima da média. A relação observada descreve a amostra, mas não prova que a temperatura causou o consumo elevado.
'''))


**Interpretação:** o filtro somente por consumo selecionou **22 registros**. Ao incluir a temperatura acima da média, o total caiu para **8 registros**, redução de **14 registros (63.64%)**.

Isso acontece porque o segundo DataFrame exige duas condições simultâneas e, portanto, contém apenas a parte dos episódios de alto consumo que também ocorreu com temperatura acima da média. A relação observada descreve a amostra, mas não prova que a temperatura causou o consumo elevado.


# Dataset 2 - Steel Industry Energy Consumption

**Situação:** a equipe de gestão de energia de uma indústria siderúrgica deseja localizar situações de consumo elevado e verificar se esses períodos coincidem com condições de carga elevada ou valores menos favoráveis de fator de potência.

A análise utiliza a amostra aleatória de 20% preparada no Orange Data Mining.

## 1. Carregamento da amostra e renomeação das colunas

O Orange acrescentou duas linhas de metadados ao CSV. Por isso, elas são ignoradas com `skiprows=[1, 2]`. Se o arquivo não estiver na pasta do notebook ou em Downloads, o código permite enviá-lo pelo Google Colab.

In [12]:
import pandas as pd
from pathlib import Path

nomes_dataset2 = [
    'Steel_industry_data_amostra_20_percent.csv'
]
pastas_busca_2 = [
    Path('/content'),
    Path.cwd(),
    Path.home() / 'Downloads',
    Path.home() / 'Documents' / 'arquivos-cp-sers'
]
arquivo = next(
    (
        pasta / nome
        for pasta in pastas_busca_2
        for nome in nomes_dataset2
        if (pasta / nome).exists()
    ),
    None
)

if arquivo is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-2.csv para o Colab antes de executar.'
    )

dados = pd.read_csv(arquivo)
inicio_dataset2 = ' '.join(
    map(str, dados.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset2 or 'meta' in inicio_dataset2:
    dados = pd.read_csv(arquivo, skiprows=[1, 2])

dados.rename(columns={
    'date': 'Data_Hora',
    'Usage_kWh': 'Consumo_kWh',
    'Leading_Current_Power_Factor': 'Fator_Potencia_Adiantado',
    'Lagging_Current_Power_Factor': 'Fator_Potencia_Atrasado',
    'Leading_Current_Reactive_Power_kVarh': 'Pot_Reativa_Adiantada_kVArh',
    'Lagging_Current_Reactive.Power_kVarh': 'Pot_Reativa_Atrasada_kVArh',
    'WeekStatus': 'Status_Semana',
    'Day_of_week': 'Dia_Semana',
    'Load_Type': 'Tipo_Carga'
}, inplace=True)

dados['Data_Hora'] = pd.to_datetime(
    dados['Data_Hora'], format='%d/%m/%Y %H:%M', errors='coerce'
)
print('Colunas após a renomeação:')
print(dados.columns.tolist())

Colunas após a renomeação:
['Data_Hora', 'Tipo_Carga', 'Dia_Semana', 'Status_Semana', 'Fator_Potencia_Adiantado', 'Fator_Potencia_Atrasado', 'Pot_Reativa_Adiantada_kVArh', 'Pot_Reativa_Atrasada_kVArh', 'Consumo_kWh']


## 2. Inspeção inicial: `head()`, `shape`, `info()` e `describe()`

In [13]:
dados.head()

,Data_Hora,Tipo_Carga,Dia_Semana,Status_Semana,Fator_Potencia_Adiantado,Fator_Potencia_Atrasado,Pot_Reativa_Adiantada_kVArh,Pot_Reativa_Atrasada_kVArh,Consumo_kWh
0,2018-04-29 07:15:00,Light_Load,Sunday,Weekend,100.00,60.20,0.0,3.82,2.88
1,2018-10-04 12:00:00,Maximum_Load,Thursday,Weekday,100.00,78.46,0.0,48.02,60.77
2,2018-01-26 11:30:00,Maximum_Load,Friday,Weekday,100.00,89.61,0.0,59.65,120.42
3,2018-06-02 14:30:00,Light_Load,Saturday,Weekend,18.53,100.00,16.6,0.00,3.13
4,2018-12-07 15:00:00,Medium_Load,Friday,Weekday,100.00,94.19,0.0,20.99,58.86


In [14]:
print(f'Dimensões da amostra: {dados.shape[0]} linhas e {dados.shape[1]} colunas')
dados.shape

Dimensões da amostra: 7008 linhas e 9 colunas


(7008, 9)

In [15]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7008 entries, 0 to 7007
Data columns (total 9 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Data_Hora                    7008 non-null   datetime64[ns]
 1   Tipo_Carga                   7008 non-null   object        
 2   Dia_Semana                   7008 non-null   object        
 3   Status_Semana                7008 non-null   object        
 4   Fator_Potencia_Adiantado     7008 non-null   float64       
 5   Fator_Potencia_Atrasado      7008 non-null   float64       
 6   Pot_Reativa_Adiantada_kVArh  7008 non-null   float64       
 7   Pot_Reativa_Atrasada_kVArh   7008 non-null   float64       
 8   Consumo_kWh                  7008 non-null   float64       
dtypes: datetime64[ns](1), float64(5), object(3)
memory usage: 492.9+ KB


In [16]:
dados.describe(include='all')

,Data_Hora,Tipo_Carga,Dia_Semana,Status_Semana,Fator_Potencia_Adiantado,Fator_Potencia_Atrasado,Pot_Reativa_Adiantada_kVArh,Pot_Reativa_Atrasada_kVArh,Consumo_kWh
count,7008,7008,7008,7008,7008.000000,7008.000000,7008.000000,7008.000000,7008.000000
unique,NaN,3,7,2,NaN,NaN,NaN,NaN,NaN
top,NaN,Light_Load,Thursday,Weekday,NaN,NaN,NaN,NaN,NaN
freq,NaN,3572,1048,5026,NaN,NaN,NaN,NaN,NaN
mean,2018-07-02 20:37:56.070205184,NaN,NaN,NaN,84.491448,80.672277,3.803711,13.206839,27.790164
min,2018-01-01 01:15:00,NaN,NaN,NaN,12.710000,37.870000,0.000000,0.000000,2.450000
25%,2018-03-31 20:52:30,NaN,NaN,NaN,99.710000,63.602500,0.000000,2.380000,3.200000
50%,2018-07-01 09:30:00,NaN,NaN,NaN,100.000000,87.990000,0.000000,5.040000,4.570000
75%,2018-10-03 07:07:30,NaN,NaN,NaN,100.000000,98.842500,1.857500,22.977500,52.107500
max,2018-12-31 23:45:00,NaN,NaN,NaN,100.000000,100.000000,27.650000,87.700000,153.140000


### Verificação complementar da qualidade e das categorias

In [17]:
print('Valores ausentes por coluna:')
display(dados.isnull().sum().to_frame('Quantidade'))

print('Categorias de Tipo_Carga:')
display(dados['Tipo_Carga'].value_counts().to_frame('Quantidade'))

print('Categorias de Status_Semana:')
display(dados['Status_Semana'].value_counts().to_frame('Quantidade'))

Valores ausentes por coluna:


,Quantidade
Data_Hora,0
Tipo_Carga,0
Dia_Semana,0
Status_Semana,0
Fator_Potencia_Adiantado,0
Fator_Potencia_Atrasado,0
Pot_Reativa_Adiantada_kVArh,0
Pot_Reativa_Atrasada_kVArh,0
Consumo_kWh,0


Categorias de Tipo_Carga:


,Quantidade
Tipo_Carga,
Light_Load,3572
Medium_Load,1937
Maximum_Load,1499


Categorias de Status_Semana:


,Quantidade
Status_Semana,
Weekday,5026
Weekend,1982


## 3. Consumo máximo e limiar de 75%

In [18]:
consumo_maximo = dados['Consumo_kWh'].max()
limiar_75 = 0.75 * consumo_maximo

print(f'Maior consumo registrado: {consumo_maximo:.2f} kWh')
print(f'Limiar de 75% do consumo máximo: {limiar_75:.2f} kWh')

Maior consumo registrado: 153.14 kWh
Limiar de 75% do consumo máximo: 114.85 kWh


## 4. DataFrame com registros acima do limiar

In [19]:
df_alto_consumo = dados[dados['Consumo_kWh'] > limiar_75].copy()
quantidade_alto_consumo = len(df_alto_consumo)
percentual_alto_consumo = quantidade_alto_consumo / len(dados) * 100

print(f'Quantidade de registros de alto consumo: {quantidade_alto_consumo}')
print(f'Percentual da amostra: {percentual_alto_consumo:.2f}%')
df_alto_consumo.sort_values('Consumo_kWh', ascending=False).head(10)

Quantidade de registros de alto consumo: 129
Percentual da amostra: 1.84%


,Data_Hora,Tipo_Carga,Dia_Semana,Status_Semana,Fator_Potencia_Adiantado,Fator_Potencia_Atrasado,Pot_Reativa_Adiantada_kVArh,Pot_Reativa_Atrasada_kVArh,Consumo_kWh
744,2018-01-15 13:45:00,Medium_Load,Monday,Weekday,100.0,90.85,0.0,70.45,153.14
4791,2018-12-19 14:15:00,Medium_Load,Wednesday,Weekday,100.0,89.45,0.0,74.56,149.18
531,2018-03-05 16:00:00,Maximum_Load,Monday,Weekday,100.0,90.16,0.0,70.49,146.88
1284,2018-01-05 16:00:00,Medium_Load,Friday,Weekday,100.0,90.94,0.0,66.92,146.34
2004,2018-01-15 13:30:00,Medium_Load,Monday,Weekday,100.0,88.54,0.0,76.79,146.27
7005,2018-02-01 12:00:00,Maximum_Load,Thursday,Weekday,100.0,87.13,0.0,81.94,145.51
6444,2018-01-18 11:15:00,Maximum_Load,Thursday,Weekday,100.0,89.39,0.0,72.18,143.93
5144,2018-01-31 14:30:00,Medium_Load,Wednesday,Weekday,100.0,93.06,0.0,56.56,143.82
825,2018-01-18 19:45:00,Maximum_Load,Thursday,Weekday,100.0,92.08,0.0,60.01,141.66
4800,2018-12-19 11:15:00,Maximum_Load,Wednesday,Weekday,100.0,89.60,0.0,69.77,140.80


## 5. Registros de alto consumo classificados como `Maximum_Load`

In [20]:
contagem_cargas_alto_consumo = df_alto_consumo['Tipo_Carga'].value_counts()
quantidade_maximum_load = (df_alto_consumo['Tipo_Carga'] == 'Maximum_Load').sum()
percentual_maximum_no_alto = quantidade_maximum_load / len(df_alto_consumo) * 100

display(contagem_cargas_alto_consumo.to_frame('Quantidade'))
print(f'Registros de alto consumo em Maximum_Load: {quantidade_maximum_load}')
print(f'Percentual dentro do DataFrame de alto consumo: {percentual_maximum_no_alto:.2f}%')

,Quantidade
Tipo_Carga,
Maximum_Load,62
Medium_Load,58
Light_Load,9


Registros de alto consumo em Maximum_Load: 62
Percentual dentro do DataFrame de alto consumo: 48.06%


**Interpretação:** a classificação `Maximum_Load` aparece em parte importante dos registros de consumo elevado, mas não em todos. Isso acontece porque o tipo de carga é uma categoria operacional, enquanto o filtro foi definido diretamente pelos valores de consumo observados na amostra.

## 6. Escolha e observação do fator de potência

Foi escolhido o fator de potência atrasado (`Fator_Potencia_Atrasado`). Para esta análise exploratória, valores abaixo de **90%** serão considerados mais baixos. O limite é coerente porque o fator de potência se aproxima de uma condição mais favorável quando está perto de 100%, e o próprio recorte de alto consumo apresenta valores aproximadamente entre 84% e 95%.

In [21]:
coluna_fator = 'Fator_Potencia_Atrasado'
limite_fator = 90.0

print('Distribuição do fator de potência atrasado na amostra completa:')
display(dados[coluna_fator].describe(percentiles=[0.10, 0.25, 0.50, 0.75, 0.90]).to_frame())

print('Distribuição do fator de potência atrasado nos registros de alto consumo:')
display(df_alto_consumo[coluna_fator].describe().to_frame())

print(f'Limite escolhido para representar valores mais baixos: {limite_fator:.2f}%')

Distribuição do fator de potência atrasado na amostra completa:


,Fator_Potencia_Atrasado
count,7008.000000
mean,80.672277
std,18.869855
min,37.870000
10%,49.987000
25%,63.602500
50%,87.990000
75%,98.842500
90%,100.000000
max,100.000000


Distribuição do fator de potência atrasado nos registros de alto consumo:


,Fator_Potencia_Atrasado
count,129.000000
mean,91.106047
std,2.534193
min,84.100000
25%,89.450000
50%,91.450000
75%,93.000000
max,95.280000


Limite escolhido para representar valores mais baixos: 90.00%


## 7. Consumo elevado e fator de potência abaixo do limite

In [22]:
df_alto_consumo_fator_baixo = dados[
    (dados['Consumo_kWh'] > limiar_75) &
    (dados[coluna_fator] < limite_fator)
].copy()

quantidade_critica = len(df_alto_consumo_fator_baixo)
percentual_amostra_critica = quantidade_critica / len(dados) * 100
percentual_dos_altos = quantidade_critica / len(df_alto_consumo) * 100

print(f'Registros com alto consumo e fator de potência abaixo de {limite_fator:.0f}%: {quantidade_critica}')
print(f'Percentual da amostra completa: {percentual_amostra_critica:.2f}%')
print(f'Percentual entre os registros de alto consumo: {percentual_dos_altos:.2f}%')
df_alto_consumo_fator_baixo.sort_values(
    ['Fator_Potencia_Atrasado', 'Consumo_kWh'],
    ascending=[True, False]
).head(10)

Registros com alto consumo e fator de potência abaixo de 90%: 42
Percentual da amostra completa: 0.60%
Percentual entre os registros de alto consumo: 32.56%


,Data_Hora,Tipo_Carga,Dia_Semana,Status_Semana,Fator_Potencia_Adiantado,Fator_Potencia_Atrasado,Pot_Reativa_Adiantada_kVArh,Pot_Reativa_Atrasada_kVArh,Consumo_kWh
4990,2018-11-14 08:45:00,Light_Load,Wednesday,Weekday,100.0,84.10,0.0,79.63,123.77
1968,2018-10-23 18:30:00,Medium_Load,Tuesday,Weekday,100.0,84.14,0.0,80.46,125.28
1634,2018-11-15 08:45:00,Light_Load,Thursday,Weekday,100.0,85.22,0.0,73.12,119.12
4367,2018-04-03 12:00:00,Maximum_Load,Tuesday,Weekday,100.0,85.34,0.0,70.16,114.88
5739,2018-01-08 08:30:00,Light_Load,Monday,Weekday,100.0,85.90,0.0,77.80,130.54
5668,2018-10-02 19:15:00,Medium_Load,Tuesday,Weekday,100.0,86.11,0.0,69.37,117.47
2433,2018-09-03 16:45:00,Maximum_Load,Monday,Weekday,100.0,86.31,0.0,69.08,118.04
977,2018-03-16 08:30:00,Light_Load,Friday,Weekday,100.0,86.36,0.0,69.26,118.66
3990,2018-06-05 11:45:00,Maximum_Load,Tuesday,Weekday,100.0,86.78,0.0,66.49,116.10
7005,2018-02-01 12:00:00,Maximum_Load,Thursday,Weekday,100.0,87.13,0.0,81.94,145.51


## 8. Interpretação final

O segundo DataFrame merece maior atenção porque reúne duas condições ao mesmo tempo: consumo superior a 75% do máximo da amostra e fator de potência atrasado inferior a 90%. Nesses períodos, a instalação apresenta demanda energética elevada junto de uma condição de fator de potência menos favorável, o que pode estar associado a maior circulação de potência reativa, perdas e uso menos eficiente da infraestrutura elétrica.

Na amostra analisada, **129 registros (1,84%)** ficaram acima do limiar de consumo. Desses, **62** foram classificados como `Maximum_Load`. O filtro combinado selecionou **42 registros**, equivalentes a **0,60% da amostra completa** e **32,56% dos registros de alto consumo**.

Esse recorte ajuda a equipe a priorizar períodos para investigação, mas não comprova sozinho a existência de falha. Uma conclusão operacional exigiria também avaliar equipamentos ligados, processo produtivo, potência reativa, histórico e critérios técnicos da instalação.

# Dataset 3 - Power Consumption of Tetouan City

**Situação:** uma equipe responsável pelo planejamento da distribuição de energia em Tétouan deseja identificar qual das três zonas apresenta o maior pico de consumo e observar as condições ambientais registradas nos momentos de maior demanda.

Esta seção utiliza exclusivamente a amostra de 15% exportada no Orange. O arquivo deve ser enviado ao Colab com o nome `DATASET-3.csv` e precisa conter as três colunas de consumo das zonas.

## 1. Carregamento da amostra e renomeação das três zonas

In [23]:
from pathlib import Path
import pandas as pd
from IPython.display import Markdown, display

nome_dataset3 = 'Tetuan_City_power_consumption_15_percent.csv'
candidatos_dataset3 = [
    Path('/content') / nome_dataset3,
    Path.cwd() / nome_dataset3,
    Path.home() / 'Downloads' / nome_dataset3,
    Path.home() / 'Documents' / 'arquivos-cp-sers' / nome_dataset3
]
arquivo_dataset3 = next(
    (caminho for caminho in candidatos_dataset3 if caminho.exists()),
    None
)

if arquivo_dataset3 is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-3.csv para o Colab antes de executar.'
    )

# O Orange pode acrescentar duas linhas de metadados ao CSV.
dados3 = pd.read_csv(arquivo_dataset3)
inicio_dataset3 = ' '.join(
    map(str, dados3.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset3 or 'meta' in inicio_dataset3:
    dados3 = pd.read_csv(arquivo_dataset3, skiprows=[1, 2])

# Algumas versões do arquivo possuem espaços duplicados em Zone 2 e Zone 3.
dados3.columns = [
    ' '.join(str(coluna).strip().split())
    for coluna in dados3.columns
]

colunas_obrigatorias_3 = [
    'Temperature',
    'Humidity',
    'Wind Speed',
    'Zone 1 Power Consumption',
    'Zone 2 Power Consumption',
    'Zone 3 Power Consumption'
]
colunas_faltantes_3 = [
    coluna for coluna in colunas_obrigatorias_3
    if coluna not in dados3.columns
]

if colunas_faltantes_3:
    raise ValueError(
        'O DATASET-3.csv está incompleto. Exporte novamente pelo Orange '
        'incluindo estas colunas: ' + ', '.join(colunas_faltantes_3)
    )

dados3.rename(columns={
    'DateTime': 'Data_Hora',
    'Temperature': 'Temperatura',
    'Humidity': 'Umidade',
    'Wind Speed': 'Velocidade_Vento',
    'Zone 1 Power Consumption': 'Consumo_Zona_1',
    'Zone 2 Power Consumption': 'Consumo_Zona_2',
    'Zone 3 Power Consumption': 'Consumo_Zona_3'
}, inplace=True)

colunas_numericas_3 = [
    'Temperatura', 'Umidade', 'Velocidade_Vento',
    'Consumo_Zona_1', 'Consumo_Zona_2', 'Consumo_Zona_3'
]
for coluna in colunas_numericas_3:
    dados3[coluna] = pd.to_numeric(dados3[coluna], errors='coerce')

if 'Data_Hora' in dados3.columns:
    dados3['Data_Hora'] = pd.to_datetime(
        dados3['Data_Hora'], errors='coerce', format='mixed'
    )

print(f'Dimensões da amostra: {dados3.shape[0]} linhas e {dados3.shape[1]} colunas')
print('Colunas após a renomeação:')
print(dados3.columns.tolist())
dados3.head()

Dimensões da amostra: 7863 linhas e 7 colunas
Colunas após a renomeação:
['Data_Hora', 'Temperatura', 'Velocidade_Vento', 'Umidade', 'Consumo_Zona_2', 'Consumo_Zona_1', 'Consumo_Zona_3']


,Data_Hora,Temperatura,Velocidade_Vento,Umidade,Consumo_Zona_2,Consumo_Zona_1,Consumo_Zona_3
0,2017-04-18 16:20:00,22.76,0.085,29.73,19396.7,32985.14532,17623.3
1,2017-12-27 17:10:00,15.92,0.085,54.44,29803.0,34737.64259,14417.3
2,2017-01-02 11:20:00,15.08,0.076,57.23,20808.5,27894.68354,16458.8
3,2017-03-19 20:20:00,12.42,0.081,72.30,23634.1,42476.93617,25089.7
4,2017-04-24 05:50:00,14.58,0.080,80.60,12006.1,21341.14101,11531.6


### Inspeção complementar da amostra

In [24]:
dados3.info()
print()
print('Valores ausentes por coluna:')
display(dados3.isnull().sum().to_frame('Quantidade'))
dados3.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7863 entries, 0 to 7862
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Data_Hora         7863 non-null   datetime64[ns]
 1   Temperatura       7863 non-null   float64       
 2   Velocidade_Vento  7863 non-null   float64       
 3   Umidade           7863 non-null   float64       
 4   Consumo_Zona_2    7863 non-null   float64       
 5   Consumo_Zona_1    7863 non-null   float64       
 6   Consumo_Zona_3    7863 non-null   float64       
dtypes: datetime64[ns](1), float64(6)
memory usage: 430.1 KB

Valores ausentes por coluna:


,Quantidade
Data_Hora,0
Temperatura,0
Velocidade_Vento,0
Umidade,0
Consumo_Zona_2,0
Consumo_Zona_1,0
Consumo_Zona_3,0


,Data_Hora,Temperatura,Velocidade_Vento,Umidade,Consumo_Zona_2,Consumo_Zona_1,Consumo_Zona_3
count,7863,7863.000000,7863.000000,7863.000000,7863.000000,7863.000000,7863.000000
mean,2017-07-02 12:29:42.449446912,18.789376,1.965602,68.400612,21060.733609,32371.549651,17845.600069
min,2017-01-01 00:40:00,3.541000,0.054000,13.340000,9436.250000,14612.658230,6211.760000
25%,2017-04-02 23:35:00,14.340000,0.078000,58.485000,17013.400000,26263.115455,13200.700000
50%,2017-07-03 20:10:00,18.690000,0.086000,70.000000,20851.600000,32249.429660,16490.300000
75%,2017-10-02 02:55:00,22.820000,4.915000,81.500000,24812.200000,37421.355555,21763.600000
max,2017-12-30 22:10:00,39.600000,6.325000,94.700000,36437.200000,51571.498340,47441.700000
std,NaN,5.864564,2.350954,15.665773,5202.357437,7134.686966,6573.203996


## 2 e 3. Consumo máximo de cada zona e identificação do maior pico

In [25]:
colunas_zonas_3 = [
    'Consumo_Zona_1',
    'Consumo_Zona_2',
    'Consumo_Zona_3'
]

maximos_zonas_3 = dados3[colunas_zonas_3].max()
zona_maior_pico_3 = maximos_zonas_3.idxmax()
maior_pico_3 = maximos_zonas_3.max()

display(maximos_zonas_3.to_frame('Consumo máximo'))
print(f'Zona com o maior pico: {zona_maior_pico_3}')
print(f'Maior pico registrado: {maior_pico_3:.2f}')

,Consumo máximo
Consumo_Zona_1,51571.49834
Consumo_Zona_2,36437.20000
Consumo_Zona_3,47441.70000


Zona com o maior pico: Consumo_Zona_1
Maior pico registrado: 51571.50


## 4 e 5. Limiar de 70%, quantidade e percentual de registros

In [26]:
limiar_70_3 = 0.70 * maior_pico_3
df_alto_consumo_3 = dados3[
    dados3[zona_maior_pico_3] > limiar_70_3
].copy()

quantidade_alto_3 = len(df_alto_consumo_3)
percentual_alto_3 = quantidade_alto_3 / len(dados3) * 100

print(f'Limiar de 70% do maior pico: {limiar_70_3:.2f}')
print(f'Registros acima do limiar: {quantidade_alto_3}')
print(f'Percentual da amostra: {percentual_alto_3:.2f}%')
df_alto_consumo_3.sort_values(
    zona_maior_pico_3, ascending=False
).head(10)

Limiar de 70% do maior pico: 36100.05
Registros acima do limiar: 2388
Percentual da amostra: 30.37%


,Data_Hora,Temperatura,Velocidade_Vento,Umidade,Consumo_Zona_2,Consumo_Zona_1,Consumo_Zona_3
6315,2017-08-17 20:10:00,25.41,0.070,57.71,34715.1,51571.49834,39778.3
3672,2017-07-24 20:20:00,28.84,4.910,32.78,34974.7,51533.82060,47405.5
3229,2017-07-29 20:10:00,28.11,4.907,46.75,34336.7,51412.62458,42055.2
6829,2017-07-27 20:20:00,27.74,4.907,61.55,34674.7,51297.80731,47435.6
7769,2017-08-08 20:30:00,26.77,4.910,84.70,35125.7,51085.63818,40314.0
988,2017-08-17 20:40:00,25.06,0.074,58.75,34163.9,50893.85128,38400.0
6642,2017-07-26 20:20:00,33.58,4.907,29.78,34651.9,50800.26578,47441.7
4445,2017-08-07 21:00:00,27.25,4.903,81.40,35152.3,50778.77913,39501.4
476,2017-07-24 20:00:00,29.39,4.909,33.38,35008.9,50755.61462,47429.6
2911,2017-08-05 20:00:00,26.49,4.905,84.00,35156.1,50682.88568,40789.5


## 6. Alto consumo e temperatura acima da média

In [27]:
temperatura_media_3 = dados3['Temperatura'].mean()
df_alto_consumo_temperatura_3 = dados3[
    (dados3[zona_maior_pico_3] > limiar_70_3) &
    (dados3['Temperatura'] > temperatura_media_3)
].copy()

quantidade_duplo_3 = len(df_alto_consumo_temperatura_3)
percentual_duplo_3 = quantidade_duplo_3 / len(dados3) * 100

print(f'Temperatura média da amostra: {temperatura_media_3:.2f} °C')
print(f'Registros com as duas condições: {quantidade_duplo_3}')
print(f'Percentual da amostra: {percentual_duplo_3:.2f}%')
df_alto_consumo_temperatura_3.head(10)

Temperatura média da amostra: 18.79 °C
Registros com as duas condições: 1597
Percentual da amostra: 20.31%


,Data_Hora,Temperatura,Velocidade_Vento,Umidade,Consumo_Zona_2,Consumo_Zona_1,Consumo_Zona_3
12,2017-06-05 23:30:00,21.03,0.068,86.50,25383.4,40993.90728,24960.0
14,2017-08-23 17:50:00,27.06,4.905,47.58,24276.2,37475.16093,25140.6
19,2017-06-22 02:40:00,19.20,4.924,65.81,24754.7,40676.02649,25095.9
21,2017-08-23 13:50:00,27.11,4.906,43.35,27416.3,40256.07103,25453.5
25,2017-07-12 09:50:00,26.92,4.909,70.20,25906.3,36511.89369,25654.9
40,2017-08-24 17:00:00,26.64,4.907,53.55,24569.0,37782.01998,22877.5
50,2017-11-13 20:10:00,19.60,0.073,67.65,31805.0,38153.84615,20483.9
52,2017-08-16 22:00:00,22.79,0.070,62.78,32449.4,48349.47836,35757.7
54,2017-08-01 09:30:00,26.60,4.910,73.10,25496.5,36369.18979,22721.0
59,2017-06-11 21:40:00,20.32,0.081,64.15,25286.1,42875.76159,27323.1


## 7. Comparação dos DataFrames e interpretação

In [28]:
reducao_3 = quantidade_alto_3 - quantidade_duplo_3
reducao_percentual_3 = (
    reducao_3 / quantidade_alto_3 * 100
    if quantidade_alto_3 > 0 else 0
)

display(Markdown(f'''
**Interpretação:** o primeiro DataFrame, que considera somente o consumo acima de 70% do maior pico da **{zona_maior_pico_3}**, possui **{quantidade_alto_3} registros**. Depois da inclusão da temperatura acima da média, o segundo DataFrame ficou com **{quantidade_duplo_3} registros**.

A segunda condição reduziu o conjunto em **{reducao_3} registros ({reducao_percentual_3:.2f}%)**, pois manteve somente os momentos de alta demanda que também ocorreram com temperatura acima da média. Essa comparação descreve a amostra analisada, mas não demonstra que a temperatura causou o aumento do consumo.
'''))


**Interpretação:** o primeiro DataFrame, que considera somente o consumo acima de 70% do maior pico da **Consumo_Zona_1**, possui **2388 registros**. Depois da inclusão da temperatura acima da média, o segundo DataFrame ficou com **1597 registros**.

A segunda condição reduziu o conjunto em **791 registros (33.12%)**, pois manteve somente os momentos de alta demanda que também ocorreram com temperatura acima da média. Essa comparação descreve a amostra analisada, mas não demonstra que a temperatura causou o aumento do consumo.


# Dataset 4 - Solar Power Generation Data

**Situação:** uma equipe de operação de uma usina fotovoltaica deseja localizar períodos de alta geração e identificar quais inversores aparecem com maior frequência nesses momentos. A análise usa o arquivo de geração da planta 1, sem junção com sensores.

## 1 e 2. Carregamento, renomeação e inspeção inicial

In [29]:
from pathlib import Path
import pandas as pd

nomes_dataset4 = [
    'DATASET-4.csv',
    'Plant_1_Generation_Data_20_percent.csv'
]
pastas_busca_4 = [
    Path('/content'),
    Path.cwd(),
    Path.home() / 'Downloads',
    Path.home() / 'Documents' / 'arquivos-cp-sers'
]
arquivo_dataset4 = next(
    (
        pasta / nome
        for pasta in pastas_busca_4
        for nome in nomes_dataset4
        if (pasta / nome).exists()
    ),
    None
)

if arquivo_dataset4 is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-4.csv para o Colab antes de executar.'
    )

dados4_brutos = pd.read_csv(arquivo_dataset4)
inicio_dataset4 = ' '.join(
    map(str, dados4_brutos.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset4 or 'meta' in inicio_dataset4:
    dados4_brutos = pd.read_csv(arquivo_dataset4, skiprows=[1, 2])

colunas_dataset4 = [
    'DATE_TIME', 'SOURCE_KEY', 'DC_POWER',
    'AC_POWER', 'DAILY_YIELD', 'TOTAL_YIELD'
]
dados4 = dados4_brutos[colunas_dataset4].copy()

dados4.rename(columns={
    'DATE_TIME': 'Data_Hora',
    'DC_POWER': 'Potencia_CC',
    'AC_POWER': 'Potencia_CA',
    'DAILY_YIELD': 'Geracao_Diaria',
    'TOTAL_YIELD': 'Geracao_Total'
}, inplace=True)

for coluna in ['Potencia_CC', 'Potencia_CA', 'Geracao_Diaria', 'Geracao_Total']:
    dados4[coluna] = pd.to_numeric(dados4[coluna], errors='coerce')
dados4.dropna(inplace=True)
dados4.reset_index(drop=True, inplace=True)

print(f'Dimensões da amostra: {dados4.shape[0]} linhas e {dados4.shape[1]} colunas')
dados4.head()

Dimensões da amostra: 13756 linhas e 6 colunas


,Data_Hora,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,Geracao_Total
0,10-06-2020 19:00,uHbuxQJl8lW7ozc,0.00,0.000,6645.000,7240508.000
1,28-05-2020 00:15,rGa61gmuvPhdLxV,0.00,0.000,0.000,7206549.000
2,17-05-2020 15:15,sjndEbLyjtCKgGv,7333.88,718.062,6442.000,7036055.000
3,01-06-2020 04:15,ih0vzX44oOqAx2f,0.00,0.000,0.000,6308232.000
4,17-05-2020 09:00,z9Y9gH1T5YWrNuG,7665.50,750.438,961.375,7021701.375


In [30]:
dados4.info()
print()
print('Valores ausentes por coluna:')
display(dados4.isnull().sum().to_frame('Quantidade'))
dados4.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13756 entries, 0 to 13755
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Data_Hora       13756 non-null  object 
 1   SOURCE_KEY      13756 non-null  object 
 2   Potencia_CC     13756 non-null  float64
 3   Potencia_CA     13756 non-null  float64
 4   Geracao_Diaria  13756 non-null  float64
 5   Geracao_Total   13756 non-null  float64
dtypes: float64(4), object(2)
memory usage: 644.9+ KB

Valores ausentes por coluna:


,Quantidade
Data_Hora,0
SOURCE_KEY,0
Potencia_CC,0
Potencia_CA,0
Geracao_Diaria,0
Geracao_Total,0


,Data_Hora,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,Geracao_Total
count,13756,13756,13756.000000,13756.000000,13756.000000,1.375600e+04
unique,3122,22,NaN,NaN,NaN,NaN
top,10-06-2020 19:00,z9Y9gH1T5YWrNuG,NaN,NaN,NaN,NaN
freq,11,672,NaN,NaN,NaN,NaN
mean,NaN,NaN,3135.848662,306.695495,3319.235216,6.978829e+06
std,NaN,NaN,4023.069619,393.124284,3142.249927,4.162164e+05
min,NaN,NaN,0.000000,0.000000,0.000000,6.183645e+06
25%,NaN,NaN,0.000000,0.000000,0.000000,6.509779e+06
50%,NaN,NaN,370.071500,35.771400,2730.425000,7.146218e+06
75%,NaN,NaN,6350.562500,622.065500,6293.782500,7.267996e+06


## 3 e 4. Potência CA máxima e limiar de 70%

In [31]:
potencia_ca_maxima = dados4['Potencia_CA'].max()
limiar_70_ca = 0.70 * potencia_ca_maxima

print(f'Maior potência CA registrada: {potencia_ca_maxima:.2f}')
print(f'Limiar de 70% da potência CA máxima: {limiar_70_ca:.2f}')

Maior potência CA registrada: 1405.59
Limiar de 70% da potência CA máxima: 983.91


## 5. Registros de alta geração

In [32]:
df_alta_geracao_4 = dados4[
    dados4['Potencia_CA'] > limiar_70_ca
].copy()

quantidade_alta_4 = len(df_alta_geracao_4)
percentual_alta_4 = quantidade_alta_4 / len(dados4) * 100

print(f'Registros acima do limiar: {quantidade_alta_4}')
print(f'Percentual da amostra: {percentual_alta_4:.2f}%')
df_alta_geracao_4.sort_values(
    'Potencia_CA', ascending=False
).head(10)

Registros acima do limiar: 1185
Percentual da amostra: 8.61%


,Data_Hora,SOURCE_KEY,Potencia_CC,Potencia_CA,Geracao_Diaria,Geracao_Total
13172,14-06-2020 14:00,adLQvlD726eNBSB,14416.1,1405.59,5471.29,6502967.286
2822,22-05-2020 11:30,adLQvlD726eNBSB,14300.3,1394.29,2760.14,6324333.143
10942,14-06-2020 14:00,ih0vzX44oOqAx2f,14126.4,1377.31,5091.29,6405609.286
1299,14-06-2020 14:00,uHbuxQJl8lW7ozc,14125.6,1377.25,5235.50,7265766.500
3323,07-06-2020 11:15,McdE0feGgRqW7Ca,14091.7,1373.93,2826.57,7333986.571
8718,30-05-2020 12:00,YxYtjZvoooNbGkE,13899.9,1355.23,2552.86,7295055.857
11606,07-06-2020 11:15,zVJPv84UY57bAof,13868.9,1352.19,2753.14,7288283.143
2654,17-06-2020 12:15,YxYtjZvoooNbGkE,13839.9,1349.40,3128.57,7422684.571
2275,07-06-2020 11:15,z9Y9gH1T5YWrNuG,13821.1,1347.56,2762.86,7180051.857
6304,30-05-2020 12:00,rGa61gmuvPhdLxV,13818.5,1347.30,2562.25,7225636.250


## 6 e 7. Frequência dos inversores e interpretação

In [33]:
frequencia_inversores_4 = df_alta_geracao_4['SOURCE_KEY'].value_counts()
inversor_frequente_4 = frequencia_inversores_4.idxmax()
quantidade_inversor_4 = frequencia_inversores_4.max()

display(frequencia_inversores_4.to_frame('Quantidade'))
print(f'Inversor mais frequente: {inversor_frequente_4}')
print(f'Ocorrências nos registros de alta geração: {quantidade_inversor_4}')

,Quantidade
SOURCE_KEY,
adLQvlD726eNBSB,68
McdE0feGgRqW7Ca,65
ZnxXDlPa8U1GXgE,63
pkci93gMrogZuBj,62
ZoEaEvLYb1n2sOq,61
1IF53ai7Xc0U56Y,59
YxYtjZvoooNbGkE,58
sjndEbLyjtCKgGv,57
z9Y9gH1T5YWrNuG,56


Inversor mais frequente: adLQvlD726eNBSB
Ocorrências nos registros de alta geração: 68


In [34]:
display(Markdown(f'''
**Interpretação:** o inversor **{inversor_frequente_4}** apareceu com maior frequência no recorte de alta geração, com **{quantidade_inversor_4} registros**. Isso permite observar sua presença relativa nos momentos acima de 70% da potência CA máxima da amostra.

O resultado, isoladamente, não demonstra que esse inversor possui desempenho superior nem que os demais apresentam falha. Uma conclusão desse tipo exigiria comparar disponibilidade, irradiância, horários de operação, quantidade total de registros e condições técnicas de cada equipamento.
'''))


**Interpretação:** o inversor **adLQvlD726eNBSB** apareceu com maior frequência no recorte de alta geração, com **68 registros**. Isso permite observar sua presença relativa nos momentos acima de 70% da potência CA máxima da amostra.

O resultado, isoladamente, não demonstra que esse inversor possui desempenho superior nem que os demais apresentam falha. Uma conclusão desse tipo exigiria comparar disponibilidade, irradiância, horários de operação, quantidade total de registros e condições técnicas de cada equipamento.


# Dataset 5 - Wind & Solar Energy Production

O arquivo está no formato longo: cada linha identifica o tipo de geração e o valor produzido. Por isso, a análise separa os registros solares e eólicos antes de calcular os máximos e os limiares.

## 1. Carregamento e inspeção da amostra fornecida

In [35]:
from pathlib import Path
import pandas as pd

nomes_dataset5 = [
    'Energy_Production_Dataset_20_percent.csv'
]
pastas_busca_5 = [
    Path('/content'),
    Path.cwd(),
    Path.home() / 'Downloads',
    Path.home() / 'Documents' / 'arquivos-cp-sers'
]
arquivo_dataset5 = next(
    (
        pasta / nome
        for pasta in pastas_busca_5
        for nome in nomes_dataset5
        if (pasta / nome).exists()
    ),
    None
)

if arquivo_dataset5 is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-5.csv para o Colab antes de executar.'
    )

dados5 = pd.read_csv(arquivo_dataset5)
inicio_dataset5 = ' '.join(
    map(str, dados5.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset5 or 'meta' in inicio_dataset5:
    dados5 = pd.read_csv(arquivo_dataset5, skiprows=[1, 2])

dados5.rename(columns={'Source': 'Tipo_Geracao'}, inplace=True)
dados5['Production'] = pd.to_numeric(
    dados5['Production'], errors='coerce'
)
for coluna in ['Day_of_Year', 'Start_Hour', 'End_Hour']:
    dados5[coluna] = pd.to_numeric(dados5[coluna], errors='coerce')

dados5.dropna(subset=['Tipo_Geracao', 'Production'], inplace=True)
dados5.reset_index(drop=True, inplace=True)

print(f'Dimensões da amostra: {dados5.shape[0]} linhas e {dados5.shape[1]} colunas')
dados5.head()

Dimensões da amostra: 10373 linhas e 9 colunas


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Production
0,12/18/2022,December,Sunday,Wind,352,22,21,Winter,11400
1,1/29/2024,January,Monday,Wind,29,15,14,Winter,7917
2,8/1/2024,August,Thursday,Solar,214,16,15,Summer,8835
3,11/10/2020,November,Tuesday,Wind,315,2,1,Fall,989
4,12/19/2022,December,Monday,Wind,353,2,1,Winter,12526


In [36]:
dados5.info()
print()
print('Valores ausentes por coluna:')
display(dados5.isnull().sum().to_frame('Quantidade'))
print()
print('Registros por tipo de geração:')
display(dados5['Tipo_Geracao'].value_counts().to_frame('Quantidade'))
dados5.describe(include='all')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10373 entries, 0 to 10372
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Date          10373 non-null  object
 1   Month_Name    10373 non-null  object
 2   Day_Name      10373 non-null  object
 3   Tipo_Geracao  10373 non-null  object
 4   Day_of_Year   10373 non-null  int64 
 5   End_Hour      10373 non-null  int64 
 6   Start_Hour    10373 non-null  int64 
 7   Season        10373 non-null  object
 8   Production    10373 non-null  int64 
dtypes: int64(4), object(5)
memory usage: 729.5+ KB

Valores ausentes por coluna:


,Quantidade
Date,0
Month_Name,0
Day_Name,0
Tipo_Geracao,0
Day_of_Year,0
End_Hour,0
Start_Hour,0
Season,0
Production,0



Registros por tipo de geração:


,Quantidade
Tipo_Geracao,
Wind,8527
Solar,1846


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Production
count,10373,10373,10373,10373,10373.000000,10373.000000,10373.000000,10373,10373.000000
unique,2150,12,7,2,NaN,NaN,NaN,4,NaN
top,1/27/2024,April,Monday,Wind,NaN,NaN,NaN,Spring,NaN
freq,12,908,1535,8527,NaN,NaN,NaN,2686,NaN
mean,NaN,NaN,NaN,NaN,181.003181,11.489733,11.530898,NaN,6193.352164
std,NaN,NaN,NaN,NaN,103.655309,6.949144,6.947598,NaN,3981.573520
min,NaN,NaN,NaN,NaN,1.000000,0.000000,0.000000,NaN,91.000000
25%,NaN,NaN,NaN,NaN,92.000000,5.000000,6.000000,NaN,3117.000000
50%,NaN,NaN,NaN,NaN,180.000000,11.000000,12.000000,NaN,5334.000000
75%,NaN,NaN,NaN,NaN,271.000000,18.000000,18.000000,NaN,8443.000000


## 2. Separação das produções solar e eólica

In [37]:
dados5_solar = dados5[
    dados5['Tipo_Geracao'].str.strip().str.lower() == 'solar'
].copy()
dados5_eolica = dados5[
    dados5['Tipo_Geracao'].str.strip().str.lower() == 'wind'
].copy()

dados5_solar.rename(
    columns={'Production': 'Geracao_Solar'}, inplace=True
)
dados5_eolica.rename(
    columns={'Production': 'Geracao_Eolica'}, inplace=True
)

print(f'Registros solares: {len(dados5_solar)}')
print(f'Registros eólicos: {len(dados5_eolica)}')
display(dados5_solar.head())
display(dados5_eolica.head())

Registros solares: 1846
Registros eólicos: 8527


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Geracao_Solar
2,8/1/2024,August,Thursday,Solar,214,16,15,Summer,8835
5,10/18/2024,October,Friday,Solar,292,14,13,Fall,4855
6,4/10/2020,April,Friday,Solar,101,18,17,Spring,3692
13,6/2/2021,June,Wednesday,Solar,153,16,15,Summer,5561
16,5/2/2021,May,Sunday,Solar,122,14,13,Spring,4993


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Geracao_Eolica
0,12/18/2022,December,Sunday,Wind,352,22,21,Winter,11400
1,1/29/2024,January,Monday,Wind,29,15,14,Winter,7917
3,11/10/2020,November,Tuesday,Wind,315,2,1,Fall,989
4,12/19/2022,December,Monday,Wind,353,2,1,Winter,12526
7,1/25/2023,January,Wednesday,Wind,25,1,0,Winter,2156


## 3. Máximos e limiares próprios de 70%

In [38]:
maximo_solar_5 = dados5_solar['Geracao_Solar'].max()
maximo_eolico_5 = dados5_eolica['Geracao_Eolica'].max()
limiar_solar_5 = 0.70 * maximo_solar_5
limiar_eolico_5 = 0.70 * maximo_eolico_5

print(f'Máximo solar: {maximo_solar_5:.2f}')
print(f'Limiar solar de 70%: {limiar_solar_5:.2f}')
print(f'Máximo eólico: {maximo_eolico_5:.2f}')
print(f'Limiar eólico de 70%: {limiar_eolico_5:.2f}')

Máximo solar: 16316.00
Limiar solar de 70%: 11421.20
Máximo eólico: 22929.00
Limiar eólico de 70%: 16050.30


## 4 e 5. DataFrames de alta geração, quantidades e percentuais

In [39]:
df_alta_solar_5 = dados5_solar[
    dados5_solar['Geracao_Solar'] > limiar_solar_5
].copy()
df_alta_eolica_5 = dados5_eolica[
    dados5_eolica['Geracao_Eolica'] > limiar_eolico_5
].copy()

quantidade_solar_5 = len(df_alta_solar_5)
quantidade_eolica_5 = len(df_alta_eolica_5)
percentual_solar_5 = quantidade_solar_5 / len(dados5_solar) * 100
percentual_eolica_5 = quantidade_eolica_5 / len(dados5_eolica) * 100

print(
    f'Alta geração solar: {quantidade_solar_5} registros '
    f'({percentual_solar_5:.2f}% dos registros solares)'
)
print(
    f'Alta geração eólica: {quantidade_eolica_5} registros '
    f'({percentual_eolica_5:.2f}% dos registros eólicos)'
)

display(
    df_alta_solar_5.sort_values(
        'Geracao_Solar', ascending=False
    ).head(10)
)
display(
    df_alta_eolica_5.sort_values(
        'Geracao_Eolica', ascending=False
    ).head(10)
)

Alta geração solar: 45 registros (2.44% dos registros solares)
Alta geração eólica: 245 registros (2.87% dos registros eólicos)


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Geracao_Solar
4700,7/16/2025,July,Wednesday,Solar,197,15,14,Summer,16316
187,7/16/2025,July,Wednesday,Solar,197,13,12,Summer,16160
7872,7/16/2025,July,Wednesday,Solar,197,16,15,Summer,16069
10130,3/6/2025,March,Thursday,Solar,65,13,12,Spring,15080
4208,7/14/2023,July,Friday,Solar,195,14,13,Summer,13917
178,3/19/2025,March,Wednesday,Solar,78,14,13,Spring,13786
2182,5/30/2023,May,Tuesday,Solar,150,13,12,Spring,13676
77,7/8/2025,July,Tuesday,Solar,189,14,13,Summer,13125
2549,7/11/2025,July,Friday,Solar,192,15,14,Summer,13109
815,7/11/2025,July,Friday,Solar,192,14,13,Summer,13057


,Date,Month_Name,Day_Name,Tipo_Geracao,Day_of_Year,End_Hour,Start_Hour,Season,Geracao_Eolica
243,11/24/2024,November,Sunday,Wind,329,13,12,Fall,22929
2561,2/26/2023,February,Sunday,Wind,57,12,11,Winter,22043
5045,3/31/2023,March,Friday,Wind,90,14,13,Spring,21624
6642,3/31/2023,March,Friday,Wind,90,15,14,Spring,21589
3257,11/24/2024,November,Sunday,Wind,329,11,10,Fall,21544
2123,3/31/2023,March,Friday,Wind,90,16,15,Spring,21172
8491,10/23/2025,October,Thursday,Wind,296,11,10,Fall,21130
8842,10/21/2023,October,Saturday,Wind,294,13,12,Fall,20931
3603,1/28/2025,January,Tuesday,Wind,28,14,13,Winter,20885
3311,10/27/2025,October,Monday,Wind,300,11,10,Fall,20785


## 6 e 7. Comparação e interpretação

In [40]:
from IPython.display import Markdown, display

if percentual_solar_5 > percentual_eolica_5:
    geracao_frequente_5 = 'solar'
elif percentual_eolica_5 > percentual_solar_5:
    geracao_frequente_5 = 'eólica'
else:
    geracao_frequente_5 = 'solar e eólica, com a mesma frequência'

display(Markdown(f'''
**Interpretação:** a geração **{geracao_frequente_5}** apareceu com maior frequência acima de 70% do seu próprio máximo. O recorte solar correspondeu a **{percentual_solar_5:.2f}% dos registros solares**, enquanto o recorte eólico correspondeu a **{percentual_eolica_5:.2f}% dos registros eólicos**.

Não seria adequado aplicar o mesmo limite numérico aos dois tipos de geração porque suas escalas são diferentes. O máximo solar foi **{maximo_solar_5:.2f}**, enquanto o máximo eólico foi **{maximo_eolico_5:.2f}**. O uso de 70% do máximo de cada tipo de geração permite uma comparação relativa mais coerente.
'''))


**Interpretação:** a geração **eólica** apareceu com maior frequência acima de 70% do seu próprio máximo. O recorte solar correspondeu a **2.44% dos registros solares**, enquanto o recorte eólico correspondeu a **2.87% dos registros eólicos**.

Não seria adequado aplicar o mesmo limite numérico aos dois tipos de geração porque suas escalas são diferentes. O máximo solar foi **16316.00**, enquanto o máximo eólico foi **22929.00**. O uso de 70% do máximo de cada tipo de geração permite uma comparação relativa mais coerente.


# Dataset 6 - Individual Household Electric Power Consumption

**Situação:** uma residência possui monitoramento elétrico detalhado e deseja identificar episódios de demanda elevada que também apresentem corrente acima do comportamento médio.

## 1. Carregamento, tratamento e organização dos atributos

In [41]:
from pathlib import Path
import pandas as pd

nomes_dataset6 = [
    'household_power_consumption_10_percent.csv'
]
pastas_busca_6 = [
    Path('/content'),
    Path.cwd(),
    Path.home() / 'Downloads',
    Path.home() / 'Documents' / 'arquivos-cp-sers'
]
arquivo_dataset6 = next(
    (
        pasta / nome
        for pasta in pastas_busca_6
        for nome in nomes_dataset6
        if (pasta / nome).exists()
    ),
    None
)

if arquivo_dataset6 is None:
    raise FileNotFoundError(
        'Envie o arquivo DATASET-6.csv para o Colab antes de executar.'
    )

colunas_originais_6 = [
    'Global_active_power',
    'Global_reactive_power',
    'Voltage',
    'Global_intensity',
    'Sub_metering_1',
    'Sub_metering_2',
    'Sub_metering_3'
]

separador_6 = ';' if arquivo_dataset6.suffix.lower() == '.txt' else None
dados6_brutos = pd.read_csv(
    arquivo_dataset6,
    sep=separador_6,
    engine='python',
    na_values=['?', '']
)
inicio_dataset6 = ' '.join(
    map(str, dados6_brutos.head(2).to_numpy().ravel())
).lower()
if 'continuous' in inicio_dataset6 or 'meta' in inicio_dataset6:
    dados6_brutos = pd.read_csv(
        arquivo_dataset6,
        sep=separador_6,
        engine='python',
        skiprows=[1, 2],
        na_values=['?', '']
    )

dados6 = dados6_brutos[colunas_originais_6].copy()
for coluna in colunas_originais_6:
    dados6[coluna] = pd.to_numeric(dados6[coluna], errors='coerce')

linhas_antes_6 = len(dados6)
dados6.dropna(inplace=True)
linhas_removidas_6 = linhas_antes_6 - len(dados6)
dados6.reset_index(drop=True, inplace=True)

dados6.rename(columns={
    'Global_active_power': 'Potencia_Ativa',
    'Global_reactive_power': 'Potencia_Reativa',
    'Voltage': 'Tensao',
    'Global_intensity': 'Corrente',
    'Sub_metering_1': 'Submedicao_1',
    'Sub_metering_2': 'Submedicao_2',
    'Sub_metering_3': 'Submedicao_3'
}, inplace=True)

print(f'Linhas incompletas removidas: {linhas_removidas_6}')
print(f'Dimensões da amostra tratada: {dados6.shape[0]} linhas e {dados6.shape[1]} colunas')
dados6.head()

Linhas incompletas removidas: 0
Dimensões da amostra tratada: 75751 linhas e 7 colunas


,Potencia_Ativa,Potencia_Reativa,Tensao,Corrente,Submedicao_1,Submedicao_2,Submedicao_3
0,2.304,0.062,240.88,10.6,19,0,18
1,2.094,0.000,238.21,8.8,0,0,18
2,1.446,0.150,241.31,6.0,0,0,19
3,1.562,0.494,238.66,6.8,0,2,0
4,0.160,0.000,240.43,1.0,0,0,0


In [42]:
dados6.info()
print()
print('Valores ausentes após o tratamento:')
display(dados6.isnull().sum().to_frame('Quantidade'))
dados6.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75751 entries, 0 to 75750
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Potencia_Ativa    75751 non-null  float64
 1   Potencia_Reativa  75751 non-null  float64
 2   Tensao            75751 non-null  float64
 3   Corrente          75751 non-null  float64
 4   Submedicao_1      75751 non-null  int64  
 5   Submedicao_2      75751 non-null  int64  
 6   Submedicao_3      75751 non-null  int64  
dtypes: float64(4), int64(3)
memory usage: 4.0 MB

Valores ausentes após o tratamento:


,Quantidade
Potencia_Ativa,0
Potencia_Reativa,0
Tensao,0
Corrente,0
Submedicao_1,0
Submedicao_2,0
Submedicao_3,0


,Potencia_Ativa,Potencia_Reativa,Tensao,Corrente,Submedicao_1,Submedicao_2,Submedicao_3
count,75751.000000,75751.000000,75751.000000,75751.000000,75751.000000,75751.000000,75751.000000
mean,1.169126,0.115994,239.847353,4.969809,1.214994,1.649483,6.032396
std,1.167936,0.109279,3.516243,4.930530,6.390772,6.826430,8.281251
min,0.084000,0.000000,224.870000,0.400000,0.000000,0.000000,0.000000
25%,0.298000,0.000000,237.540000,1.400000,0.000000,0.000000,0.000000
50%,0.616000,0.096000,240.160000,2.800000,0.000000,0.000000,0.000000
75%,1.602000,0.182000,242.300000,6.800000,0.000000,1.000000,17.000000
max,9.922000,1.192000,251.770000,43.200000,76.000000,76.000000,30.000000


## 2 e 3. Potência ativa máxima, limiar de 75% e primeiro DataFrame

In [43]:
potencia_ativa_maxima_6 = dados6['Potencia_Ativa'].max()
limiar_75_6 = 0.75 * potencia_ativa_maxima_6
df_alta_potencia_6 = dados6[
    dados6['Potencia_Ativa'] > limiar_75_6
].copy()

print(f'Potência ativa máxima: {potencia_ativa_maxima_6:.3f} kW')
print(f'Limiar de 75%: {limiar_75_6:.3f} kW')
df_alta_potencia_6.sort_values(
    'Potencia_Ativa', ascending=False
).head(10)

Potência ativa máxima: 9.922 kW
Limiar de 75%: 7.442 kW


,Potencia_Ativa,Potencia_Reativa,Tensao,Corrente,Submedicao_1,Submedicao_2,Submedicao_3
53731,9.922,0.414,230.17,43.2,34,70,16
29223,9.254,0.000,228.56,40.4,71,0,16
38984,9.012,0.286,228.01,39.4,35,0,16
64748,8.974,0.576,230.04,39.2,36,34,0
21851,8.944,0.502,229.75,39.0,36,68,17
64349,8.632,0.288,234.02,36.8,38,1,0
66578,8.526,0.406,231.89,36.8,36,34,0
71994,8.392,0.186,236.41,35.4,37,73,17
51475,8.362,0.206,228.52,36.6,76,0,16
65034,8.284,0.064,238.17,34.8,38,0,17


## 4. Quantidade e percentual dos registros selecionados

In [44]:
quantidade_alta_6 = len(df_alta_potencia_6)
percentual_alta_6 = quantidade_alta_6 / len(dados6) * 100

print(f'Registros acima do limiar: {quantidade_alta_6}')
print(f'Percentual da amostra: {percentual_alta_6:.4f}%')

Registros acima do limiar: 52
Percentual da amostra: 0.0686%


## 5 e 6. Corrente média e segundo DataFrame

In [45]:
corrente_media_6 = dados6['Corrente'].mean()
df_alta_potencia_corrente_6 = dados6[
    (dados6['Potencia_Ativa'] > limiar_75_6) &
    (dados6['Corrente'] > corrente_media_6)
].copy()

quantidade_duplo_6 = len(df_alta_potencia_corrente_6)
percentual_duplo_6 = quantidade_duplo_6 / len(dados6) * 100

print(f'Corrente média da amostra: {corrente_media_6:.3f} A')
print(f'Registros com as duas condições: {quantidade_duplo_6}')
print(f'Percentual da amostra: {percentual_duplo_6:.4f}%')
df_alta_potencia_corrente_6.head(10)

Corrente média da amostra: 4.970 A
Registros com as duas condições: 52
Percentual da amostra: 0.0686%


,Potencia_Ativa,Potencia_Reativa,Tensao,Corrente,Submedicao_1,Submedicao_2,Submedicao_3
3314,7.568,0.422,230.46,32.8,36,34,16
3923,7.884,0.088,233.98,33.6,36,72,17
5520,7.508,0.000,228.89,32.8,35,68,16
10221,7.998,0.424,232.22,34.4,37,36,17
11751,7.938,0.320,232.30,34.2,37,36,16
11882,7.612,0.282,237.46,32.0,38,1,17
14744,7.688,0.488,241.88,31.8,1,75,18
16231,7.562,0.190,235.92,32.0,37,61,18
16455,7.490,0.574,235.39,33.0,40,28,18
16683,7.906,0.362,229.44,34.4,36,35,17


## 7. Comparação e interpretação

In [46]:
reducao_6 = quantidade_alta_6 - quantidade_duplo_6

if reducao_6 == 0:
    efeito_corrente_6 = (
        'não reduziu a quantidade de registros, pois todos os episódios acima '
        'do limiar de potência também apresentaram corrente acima da média'
    )
else:
    efeito_corrente_6 = (
        f'reduziu a quantidade em {reducao_6} registros, mantendo somente os '
        'episódios que também apresentaram corrente acima da média'
    )

display(Markdown(f'''
**Interpretação:** o primeiro DataFrame possui **{quantidade_alta_6} registros**, e o segundo possui **{quantidade_duplo_6} registros**. A inclusão da corrente como segunda condição **{efeito_corrente_6}**.

Esse comportamento é coerente com a relação entre potência ativa e intensidade de corrente na instalação. Ainda assim, o resultado descreve a amostra analisada e deve ser interpretado junto de tensão, potência reativa e características dos equipamentos.
'''))


**Interpretação:** o primeiro DataFrame possui **52 registros**, e o segundo possui **52 registros**. A inclusão da corrente como segunda condição **não reduziu a quantidade de registros, pois todos os episódios acima do limiar de potência também apresentaram corrente acima da média**.

Esse comportamento é coerente com a relação entre potência ativa e intensidade de corrente na instalação. Ainda assim, o resultado descreve a amostra analisada e deve ser interpretado junto de tensão, potência reativa e características dos equipamentos.
